libbpf-tools/biolatency.c

```diff
@@ -0,0 +1,351 @@
+		printf("DrvIn");
+		break;
+	case REQ_OP_DRV_OUT:
+		printf("DrvOut");
+		break;
+	default:
+		printf("Unknown");
+		break;
+	}
+}
+
+static int print_log2_hists(int fd)
+{
+	struct hist_key lookup_key = { .cmd_flags = -1 }, next_key;
+	char *units = env.milliseconds ? "msecs" : "usecs";
+	struct hist hist;
+	int err;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		if (env.per_disk)
+			printf("\ndisk = %s\t", next_key.disk[0] != '\0' ?
+				next_key.disk : "unnamed");
+		if (env.per_flag)
+			print_cmd_flags(next_key.cmd_flags);
+		printf("\n");
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		lookup_key = next_key;
+	}
+
+	lookup_key.cmd_flags = -1;
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup hist : %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct biolatency_bpf *obj;
+	struct tm *tm;
+	char ts[32];
+	int fd, err;
+	time_t t;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = biolatency_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF ojbect\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	if (env.disk)
+		strncpy((char*)obj->rodata->targ_disk, env.disk, env.disk_len);
+	obj->rodata->targ_per_disk = env.per_disk;
+	obj->rodata->targ_per_flag = env.per_flag;
+	obj->rodata->targ_ms = env.milliseconds;
+	obj->rodata->targ_queued = env.queued;
+
+	err = biolatency_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	if (env.queued) {
+		obj->links.tp_btf__block_rq_insert =
+			bpf_program__attach(obj->progs.tp_btf__block_rq_insert);
+		err = libbpf_get_error(obj->links.tp_btf__block_rq_insert);
+		if (err) {
+			fprintf(stderr, "failed to attach: %s\n", strerror(-err));
+			goto cleanup;
+		}
+	}
+	obj->links.tp_btf__block_rq_issue =
+		bpf_program__attach(obj->progs.tp_btf__block_rq_issue);
+	err = libbpf_get_error(obj->links.tp_btf__block_rq_issue);
+	if (err) {
+		fprintf(stderr, "failed to attach: %s\n", strerror(-err));
+		goto cleanup;
+	}
+	obj->links.tp_btf__block_rq_complete =
+		bpf_program__attach(obj->progs.tp_btf__block_rq_complete);
+	err = libbpf_get_error(obj->links.tp_btf__block_rq_complete);
+	if (err) {
+		fprintf(stderr, "failed to attach: %s\n", strerror(-err));
+		goto cleanup;
+	}
+
+	fd = bpf_map__fd(obj->maps.hists);
```

Comment:
you use it in one place below, I'd just do it there without extra `fd` variable. It's extremely cheap function.

Reply:
Fixed.